# HG changeset patch
# User David Sterba <dsterba@suse.cz>
# Date 1337866493 -7200
#      Thu May 24 15:34:53 2012 +0200
# Branch gentoo-1.7
# Node ID b02d21b36b0edd4fea57c434fcae539f0ff08372
# Parent  9366196c22a099decaa5ca9edc97da1e0ceb5a6b
[PATCH] feature: limit-current-thread
From 658d7f5742849620f6535d5c191d204abe05b23d Mon Sep 17 00:00:00 2001
Add command to limit view on the thread from any of its message.
Predefined on command <ESC>L, though the lowercase 'l' would be better
(less keystrokes), but it's already occupied by a not-so-useful command
"show limit".

Tested on large folders, no performance problems observed.

Search in current thread is not implemented.

Signed-off-by: David Sterba <dsterba@suse.cz>
---
 OPS         |  1 +
 curs_main.c |  4 +++-
 functions.h |  1 +
 mutt_menu.h |  2 ++
 pattern.c   | 68 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 5 files changed, 75 insertions(+), 1 deletion(-)

diff -r 9366196c22a0 -r b02d21b36b0e OPS
--- a/OPS	Sat Jan 30 16:27:19 2016 +0000
+++ b/OPS	Thu May 24 15:34:53 2012 +0200
@@ -178,6 +178,7 @@ OP_VERSION "show the Mutt version number
 OP_VIEW_ATTACH "view attachment using mailcap entry if necessary"
 OP_VIEW_ATTACHMENTS "show MIME attachments"
 OP_WHAT_KEY "display the keycode for a key press"
+OP_LIMIT_CURRENT_THREAD "limit view to current thread"
 OP_MAIN_SHOW_LIMIT "show currently active limit pattern"
 OP_MAIN_COLLAPSE_THREAD "collapse/uncollapse current thread"
 OP_MAIN_COLLAPSE_ALL "collapse/uncollapse all threads"
diff -r 9366196c22a0 -r b02d21b36b0e PATCHES
--- a/PATCHES	Sat Jan 30 16:27:19 2016 +0000
+++ b/PATCHES	Thu May 24 15:34:53 2012 +0200
@@ -1,3 +1,4 @@
+patch-limit-current-thread-neomutt
 patch-quasi-delete-neomutt
 patch-tls-sni-neomutt
 patch-index-color-neomutt
diff -r 9366196c22a0 -r b02d21b36b0e curs_main.c
--- a/curs_main.c	Sat Jan 30 16:27:19 2016 +0000
+++ b/curs_main.c	Thu May 24 15:34:53 2012 +0200
@@ -1040,12 +1040,14 @@ int mutt_index_menu (void)
 	}
         break;
 
+      case OP_LIMIT_CURRENT_THREAD:
       case OP_MAIN_LIMIT:
 
 	CHECK_IN_MAILBOX;
 	menu->oldcurrent = (Context->vcount && menu->current >= 0 && menu->current < Context->vcount) ?
 		CURHDR->index : -1;
-	if (mutt_pattern_func (MUTT_LIMIT, _("Limit to messages matching: ")) == 0)
+	if (((op == OP_LIMIT_CURRENT_THREAD) && mutt_limit_current_thread(CURHDR))
+	    || ((op == OP_MAIN_LIMIT) && (mutt_pattern_func (MUTT_LIMIT, _("Limit to messages matching: ")) == 0)))
 	{
 	  if (menu->oldcurrent >= 0)
 	  {
diff -r 9366196c22a0 -r b02d21b36b0e functions.h
--- a/functions.h	Sat Jan 30 16:27:19 2016 +0000
+++ b/functions.h	Thu May 24 15:34:53 2012 +0200
@@ -112,6 +112,7 @@ const struct binding_t OpMain[] = { /* m
   { "next-undeleted",		OP_MAIN_NEXT_UNDELETED,		"j" },
   { "previous-undeleted",	OP_MAIN_PREV_UNDELETED,		"k" },
   { "limit",			OP_MAIN_LIMIT,			"l" },
+  { "limit-current-thread",	OP_LIMIT_CURRENT_THREAD,	NULL },
   { "link-threads",		OP_MAIN_LINK_THREADS,		"&" },
   { "list-reply",		OP_LIST_REPLY,			"L" },
   { "mail",			OP_MAIL,			"m" },
diff -r 9366196c22a0 -r b02d21b36b0e mutt_menu.h
--- a/mutt_menu.h	Sat Jan 30 16:27:19 2016 +0000
+++ b/mutt_menu.h	Thu May 24 15:34:53 2012 +0200
@@ -126,4 +126,6 @@ int mutt_menuLoop (MUTTMENU *);
 void index_make_entry (char *, size_t, struct menu_t *, int);
 int index_color (int);
 
+int mutt_limit_current_thread (HEADER *h);
+
 #endif /* _MUTT_MENU_H_ */
diff -r 9366196c22a0 -r b02d21b36b0e pattern.c
--- a/pattern.c	Sat Jan 30 16:27:19 2016 +0000
+++ b/pattern.c	Thu May 24 15:34:53 2012 +0200
@@ -1294,6 +1294,74 @@ void mutt_check_simple (char *s, size_t 
   }
 }
 
+/**
+ * top_of_thread - Find the first email in the current thread
+ * @h: Header of current email
+ *
+ * Returns:
+ *	THREAD*: success, email found
+ *	NULL:    on error
+ */
+static THREAD *
+top_of_thread (HEADER *h)
+{
+	THREAD *t;
+
+	if (!h)
+		return NULL;
+
+	t = h->thread;
+
+	while (t && t->parent)
+		t = t->parent;
+
+	return t;
+}
+
+/**
+ * mutt_limit_current_thread - Limit the email view to the current thread
+ * @h: Header of current email
+ *
+ * Returns:
+ *	1: Success
+ *	0: Failure
+ */
+int
+mutt_limit_current_thread (HEADER *h)
+{
+	int i;
+	THREAD *me;
+
+	if (!h)
+		return 0;
+
+	me = top_of_thread (h);
+	if (!me)
+		return 0;
+
+	Context->vcount    = 0;
+	Context->vsize     = 0;
+	Context->collapsed = 0;
+
+	for (i = 0; i < Context->msgcount; i++) {
+		Context->hdrs[i]->virtual    = -1;
+		Context->hdrs[i]->limited    = 0;
+		Context->hdrs[i]->collapsed  = 0;
+		Context->hdrs[i]->num_hidden = 0;
+
+		if (top_of_thread (Context->hdrs[i]) == me) {
+			BODY *body = Context->hdrs[i]->content;
+
+			Context->hdrs[i]->virtual = Context->vcount;
+			Context->hdrs[i]->limited = 1;
+			Context->v2r[Context->vcount] = i;
+			Context->vcount++;
+			Context->vsize += (body->length + body->offset - body->hdr_offset);
+		}
+	}
+	return 1;
+}
+
 int mutt_pattern_func (int op, char *prompt)
 {
   pattern_t *pat;
